lib/vmfloaty/abs.rb

```diff
@@ -0,0 +1,257 @@
+  end
+
+  # Retrieve an OS from ABS.
+  def self.retrieve(verbose, os_types, token, url, user)
+    #
+    # Contents of post must be like:
+    #
+    # {
+    #   "resources": {
+    #     "centos-7-i386": 1,
+    #     "ubuntu-1404-x86_64": 2
+    #   },
+    #   "job": {
+    #     "id": "12345",
+    #     "tags": {
+    #       "user": "username",
+    #     }
+    #   }
+    # }
+
+    conn = Http.get_conn(verbose, url)
+    conn.headers['X-AUTH-TOKEN'] = token if token
+
+    saved_job_id = DateTime.now.strftime('%Q')
+
+    req_obj = {
+      :resources => os_types,
+      :job       => {
+        :id   => saved_job_id,
+        :tags => {
+          :user => user,
+        },
+      },
+    }
+
+    puts "Posting to ABS #{req_obj.to_json}" if verbose
+
+    # os_string = os_type.map { |os, num| Array(os) * num }.flatten.join('+')
+    # raise MissingParamError, 'No operating systems provided to obtain.' if os_string.empty?
+    puts "Requesting VMs with job_id: #{saved_job_id}.  Will retry for up to an hour."
+    res = conn.post 'request', req_obj.to_json
+
+    retries = 360
+
+    raise AuthError, "HTTP #{res.status}: The token provided could not authenticate to the pooler.\n#{res_body}" if res.status == 401
+
+    (1..retries).each do |i|
+      queue_place, res_body = check_queue(conn, saved_job_id, req_obj)
+      return translated(res_body) if res_body
+
+      puts "Waiting 10 seconds to check if ABS request has been filled.  Queue Position: #{queue_place}... (x#{i})"
+      sleep(10)
+    end
+    nil
+  end
+
+  #
+  # We should fix the ABS API to be more like the vmpooler or nspooler api, but for now
+  #
+  def self.translated(res_body)
+    vmpooler_formatted_body = {}
+
+    res_body.each do |host|
+      if vmpooler_formatted_body[host['type']] && vmpooler_formatted_body[host['type']]['hostname'].class == Array
+        vmpooler_formatted_body[host['type']]['hostname'] << host['hostname']
+      else
+        vmpooler_formatted_body[host['type']] = { 'hostname' => [host['hostname']] }
+      end
+    end
+    vmpooler_formatted_body['ok'] = true
+
+    vmpooler_formatted_body
+  end
+
+  def self.check_queue(conn, job_id, req_obj)
+    queue_info_res = conn.get "status/queue/info/#{job_id}"
+    queue_info = JSON.parse(queue_info_res.body)
+
+    res = conn.post 'request', req_obj.to_json
+
+    unless res.body.empty?
+      res_body = JSON.parse(res.body)
+      return queue_info['queue_place'], res_body
+    end
+    [queue_info['queue_place'], nil]
+  end
+
+  def self.snapshot(_verbose, _url, _hostname, _token)
+    puts "Can't snapshot with ABS, use '--service vmpooler' (even for vms checked out with ABS)"
+  end
+
+  def self.status(verbose, url)
+    conn = Http.get_conn(verbose, url)
+
+    res = conn.get 'status'
+    JSON.parse(res.body)
+  end
+
+  def self.summary(verbose, url)
+    conn = Http.get_conn(verbose, url)
+
+    res = conn.get 'summary'
+    JSON.parse(res.body)
```

Comment:
Heh, this just returns `OK` which isn't valid JSON.
```
➜  vmfloaty git:(add_abs_vm_get) be bin/floaty status --service abs

error: 765: unexpected token at 'OK'. Use --trace to view backtrace
```

Reply:
I hadn't done status yet, but since it seems relatively easy I just pushed up a change with the status command fixed.